### ISSUE TITLE
ZAP : HTTP/2 : http2 requests are downgraded to 1.1 is TSL APN 1.1 is enabled

### ISSUE THREAD

**Issue by Jegeva** [zaproxy/zaproxy#7837]
### Describe the bug

If, in the main proxy properties, HTTP/1.1 is enabled, HTTP/2 requests are downgraded to HTTP/1.1

### Steps to reproduce the behavior

Enable image requests displaying in display -> process images in HTTP...

in the main proxy properties, HTTP/1.1 is enabled
go to http://www.http2demo.io/
click "run http/2 test"
check that the requests are actually over HTTP/1.1
 
in the main proxy properties, HTTP/1.1 is disabled
go to http://www.http2demo.io/
click "run http/2 test"
check that the requests are actually over HTTP/2

### Expected behavior

If the server is http/2 capable, zap should allow http/2 even is 1.1 is supported in the preferences

### Software versions

zap D-2023-04-17 - 20230421 - 143242

### Would you like to help fix this issue?

- [ ] Yes

**Comment by thc202** [comment]
How are you telling they are downgraded? ZAP will not force the client to a version if you have both versions enabled.

**Comment by Jegeva** [comment]
> How are you telling they are downgraded? ZAP will not force the client to a version if you have both versions enabled.

Because when 1.1 is enabled, i see requests issued in the browser being 1.1, 2 if 1.1 is disabled.
is it really downgraded by zap or is the zap proxy capabilities evaluated by the browser as "not http/2 capable", i don't really know but the http/2 requests exiting as 1.1

**Comment by thc202** [comment]
What is the behaviour that you expect when the client chooses 1.1 over 2?

**Comment by Jegeva** [comment]
> What is the behaviour that you expect when the client chooses 1.1 over 2?

if the client chooses 1.1, zaproxy should do the usual connect/1.1 proxy behaviour.

but if both options are selected zap should  

If the server is HTTP/2 capable, zaproxy should advertize itself as HTTP/2 capable
If the server is only HTTP/1.1 capable, zaproxy should advertize itself as HTTP/1.1 capable

the "advertize" is the delicate part, i don't know how the browsers are checking if a proxy can do http/2 but in my case (firefox) there is a behavioral difference when both options are selected that prevents FF from using it as http/2 proxy.

**Comment by thc202** [comment]
> if the client chooses 1.1, zaproxy should do the usual connect/1.1 proxy behaviour.

That's what's happening. (Though worth noting we are preferring 1.1 over 2 when negotiating, for compatibility.)

> but if both options are selected zap should

That would be a new feature.

**Comment by Jegeva** [comment]
> That would be a new feature.

I'm not sure, from my point of view (it is an opinion), it is a bug.
I think that, if the site offers http/2 that should be reflected by zap.
The "natural" site's behavior should be reflected as much as possible with the default set of options in force (ie 1.1 and 2 selected in TLS ALPN). 

I only discovered that because on a test i was doing (can't tell for obvious reasons) an endpoint was emitting 403 if the request wasn't coming over http/2. I only discovered this weird quirk in behavior after doing random testing of http/2 related options and juggling between FF, ZAP and chromium. A random user cannot be expected to go down this rabbithole to understand why http/2 isn't working "as expected".

**Comment by thc202** [comment]
That assumes too much of what will happen or not, ZAP allows to rewrite the requests or override the responses, the connection might not happen at all to the server expected by the client.

It's not a weird quirk, that's how ALPN works, the client is negotiating with ZAP (proxy) not the server. And why ZAP allows you to force one version or the other, if the client does not choose the one you want/expect.

**Comment by Jegeva** [comment]
Agreed, but the default zap behavior being to not reflect the natural behavior of the target server is unexpected to say the least. from a usability standpoint it's not really natural.

**Comment by thc202** [comment]
That ZAP is in the way of the server is known, same happens with other "connection properties" (e.g. TLS versions, server certificate…).
